Fix #16756: Add new android build handler

vojtechjelinek opened the pull request:
## Overview

1. This PR fixes #16756
2. This PR does the following: Add handler that provides data to Android build process.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

vojtechjelinek committed:
Add new android build handler

vojtechjelinek committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-new-android-handler

vojtechjelinek committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-new-android-handler

vojtechjelinek committed:
Add features and fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-new-android-handler

vojtechjelinek committed:
Fix tests

BenHenning commented on core/controllers/android.py:
```diff
@@ -343,3 +348,118 @@ def _create_dummy_skill(
+    # because the actual usage of 'secret' lies within the
+    # 'is_from_oppia_android_build' decorator, and here we are getting 'secret'
+    # because the decorator always passes every url_path_args to HTTP methods.
+    @acl_decorators.is_from_oppia_android_build
```
Quick question before I delve into this PR further: how does this work locally? When I tried running the web server locally and calling the endpoint with curl it failed to resolve any secrets (I had to disable the secret check in order to test the endpoints locally).

vojtechjelinek replied:
Ah, right, I need to write secrets emulator for local usage. I will do that today or tomorrow.

vojtechjelinek replied:
@BenHenning Currently, we set the secrets only using Google Cloud Console, we only fetch them using code in Oppia. I think that locally we would have an optional `--secrets` flag for the start script that would accept a dictionary of secrets to be set when starting the local server. These would be stored in an environment variable. Does that make sense? I think this is the fastest way to do this, it might be a bit annoying that you need to restart the server to reset the secrets but I think that is not that bad for this use case.

BenHenning replied:
Ah, meant to say this earlier: I really like the idea of using a CLI argument, but it seems (per my other comments & edits to the branch) that this will unfortunately not be an option.

BenHenning reviewed (changes requested):
Thanks @vojtechjelinek. I have one top-level question before I review the PR at length.

Note that I'm also verifying that this integrates correctly with what we need on Android (while also writing that code out since it doesn't exist in the public Oppia Android repo yet), so I'll probably wait to approve until after we can verify that all the data is present and works correctly on one of the test servers.

vojtechjelinek committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-new-android-handler

vojtechjelinek committed:
Add local secrets

vojtechjelinek committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-new-android-handler

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-new-android-handler

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix linting

vojtechjelinek commented:
@BenHenning Feel free to suggest your changes I can handle backend tests etc.

BenHenning commented:
Update: My PR still requires a few more days before I can be confident that my local changes to this PR are sufficient & correct. Will follow up with a PR into the PR's feature branch once I get to that stage.

BenHenning committed:
Various improvements.

Specifically:
- Fixes the Android generation controller to always recreate the topic
(in case it's already loaded)
- Adds Brazilian Portuguese translations to the Android generation
controller (specifically for the loaded Android exploration)
- Adds text content to the generated revision card that links to the
generated concept card (so that it can be accessed)
- Adds the generated topic from the Android generation controller to
all available classrooms once generated
- Updates the Android export controller to export both exploration
translations and classrooms (the latter required updating Classroom to
support to_dict functionality)
- Updates Exploration's to_dict return value to include the
exploration's version (both for consistency with other structures and
because this is required by Android)

BenHenning committed:
Merge branch 'add-new-android-handler' of https://github.com/vojtechjelinek/oppia into adjustments-to-new-android-controller

Conflicts:
	core/controllers/android.py

BenHenning committed:
Merge branch 'develop' into add-new-android-handler

BenHenning committed:
Merge branch 'add-new-android-handler' of https://github.com/vojtechjelinek/oppia into adjustments-to-new-android-controller

BenHenning commented:
Hey @vojtechjelinek. I'm trying to get my local changes set up in a fork PR for you to look at & integrate in this branch. Unfortunately, I've run into some problems running the local dev server that seem to have originated sometime after commit db4e538d672517e5092be8371d056309482d97a8 (this commit works) on this branch (I haven't root caused it both because of my lack of familiarity with web's startup process and the length of time it takes to start the dev server). I've verified that this issue happens specifically on this branch without my changes, so as far as I can tell it isn't caused by my local work.

The issue is that starting the server in dev mode (i.e. just ``python -m scripts.start``) hangs at ``Stopping Webpack Compiler`` shortly after the webpack compilation finishes. I've left this for >10 hours in case it was some sort of delay, and it doesn't ever resolve. It happens from a clean-built run on the latest version of this branch. I was wondering if it was caused by something in develop (hence why I updated the branch to the latest), but the latest develop by itself does not have this issue.

Edit: Also to clarify, I tried inspecting the entirety of the web start-up log and it doesn't show any noteworthy errors as far as I can tell, but the log is pretty long (~4k lines) so it's hard for me to know what to look for.

Any ideas?

vojtechjelinek commented:
@BenHenning Interesting I will look into it, I had a similar error but I think it was only on my work laptop and I didn't really have time to investigate it further.

BenHenning commented on scripts/servers.py:
```diff
@@ -193,6 +197,8 @@ def managed_dev_appserver(
         '--dev_appserver_log_level', log_level,
         app_yaml_path
     ]
+    env = {} if env is None else env
```
It seems like this is the line causing the problem mentioned here: https://github.com/oppia/oppia/pull/17192#issuecomment-1445219552. Not sure why yet.

BenHenning replied:
This is probably from this behavior for ``Popen``:

> If env is not None, it must be a mapping that defines the environment variables for the new process; these are used instead of the default behavior of inheriting the current process’ environment. This mapping can be str to str on any platform or bytes to bytes on POSIX platforms much like [os.environ](https://docs.python.org/3/library/os.html#os.environ) or [os.environb](https://docs.python.org/3/library/os.html#os.environb).

per: https://docs.python.org/3/library/subprocess.html#popen-constructor.

We probably need to default to the OS's environment table and augment it rather than replace it.

BenHenning replied:
I also found that the ``json.loads`` line in start.py wouldn't work if no secrets were passed in, so that will need to be fixed, too.

I'll add these fixes to my PR that I'll be sending out to make things a bit easier.

BenHenning commented on core/platform/models.py:
```diff
@@ -392,8 +392,12 @@ def import_secrets_services(cls) -> ModuleType:
-        from core.platform.secrets import cloud_secrets_services
-        return cloud_secrets_services
+        if constants.EMULATOR_MODE:
+            from core.platform.secrets import cloud_secrets_services
```
Will cloud secrets ever work in a local environment? Also, is this logic correct--shouldn't we be importing cloud secrets by default rather than dev mode? I'm assuming production has both DEV_MODE and EMULATOR_MODE disabled.

BenHenning replied:
This has been changed to gate on dev_mode, instead.

BenHenning committed:
Fixes some issues after integration.

Specific problems addressed:
- Classrooms don't use models today, so switched the new Android handler
over to falling back to the config domain so things will actually load
correctly.
- Fixed an issue with env being overwritten (it's now properly
appended). However, env vars have to be defined in app.yaml, so the new
Android secret was moved over there. There's potential cleanup for a
subsequent commit.
- Some correct changes were removed in the merge commit; these have been
reverted.
- The logic for using the dev secrets manager was using emulator
settings as the indicator rather than dev mode; this has been changed.

BenHenning commented:
Okay, I've been able to isolate the core issue, I think, and push my changes here: https://github.com/vojtechjelinek/oppia/pull/11. PTAL @vojtechjelinek.

Other than those changes, @seanlip is concerned about potential QPS from the download script aggressively looking for structures, translations, etc. in short succession. He suggests that we add multi-fetch functionality at the outset so that the script can utilize it if needed.

Separately from that, I'm wondering if passing the secret via HTTP arguments might be a better approach than using a path (since it'll be a bit more hidden from things like request logs and generally be "less obvious").

Do you have any thoughts and/or concerns with the above suggestions?

vojtechjelinek committed:
Merge pull request #11 from BenHenning/adjustments-to-new-android-controller

vojtechjelinek committed:
Fix linting

vojtechjelinek committed:
Fix some tests

vojtechjelinek committed:
git pushMerge branch 'develop' of https://github.com/oppia/oppia into add-new-android-handler

vojtechjelinek committed:
Fix linting

vojtechjelinek committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-new-android-handler

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-new-android-handler

vojtechjelinek committed:
Fix linting

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek commented:
@paprajapati9 PTAL

gp201 commented:
@vojtechjelinek please assign me after the backend tests have been fixed. Thanks!

vojtechjelinek committed:
Fix tests

seanlip commented on app_dev.yaml:
```diff
@@ -142,3 +142,4 @@ env_variables:
   DATASTORE_HOST: "http://localhost:8089"
   DATASTORE_PROJECT_ID: "dev-project-id"
   DATASTORE_USE_PROJECT_ID_AS_APP_ID: "true"
+  SECRETS: "{\"ANDROID_BUILD_SECRET\": \"android-dev-secret\"}"
```
Is it possible to make the name and the (I assume) default value match?

vojtechjelinek replied:
@BenHenning Is this actually needed anymore? And if yes is it fine to change the value to "ANDROID_BUILD_SECRET" or something similar?

BenHenning replied:
Yep, we can't use the scripts argument because it doesn't work correctly (and causes run failures in the dev server). We can change this value to anything so long as it's kept relatively consistent (so that we can update the corresponding downloads script).

vojtechjelinek replied:
Okay, changed to `android-build-secre`

seanlip reviewed (commented):
Generally looks fine for codeowners, just one comment.

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Fix test

vojtechjelinek committed:
Fix test

JayVivarekar commented on core/feconf.py:
```diff
@@ -1643,6 +1643,7 @@ def get_empty_ratings() -> Dict[str, int]:
 # The task entry ID template used by the task entry model.
 TASK_ENTRY_ID_TEMPLATE = '%s.%s.%d.%s.%s.%s'
 
+
```
Why is this added?

vojtechjelinek replied:
Removed.

JayVivarekar reviewed (changes requested):
LGTM for codeowner files. One nit

gp201 reviewed (approved):
LGTM for codeowner files

aasiffaizal commented on core/controllers/android_test.py:
```diff
@@ -0,0 +1,320 @@
+
+    def test_initialize_in_production_raises_exception(self) -> None:
+        prod_mode_swap = self.swap(constants, 'DEV_MODE', False)
+        assert_raises_regexp_context_manager = self.assertRaisesRegex(
+            Exception, 'Cannot load new structures data in production.')
```
```suggestion
        assert_raises_regexp_context_manager = self.assertRaisesRegex(
            Exception, 'Cannot load new structures data in production.'
        )
```

vojtechjelinek replied:
Done.

aasiffaizal commented on core/controllers/android_test.py:
```diff
@@ -0,0 +1,320 @@
+        assert_raises_regexp_context_manager = self.assertRaisesRegex(
+            Exception, 'Cannot load new structures data in production.')
+        with assert_raises_regexp_context_manager, prod_mode_swap:
+            self.post_json(
+                '/initialize_android_test_data', {}, use_payload=False,
+                csrf_token=None)
```
```suggestion
            self.post_json(
                '/initialize_android_test_data', {}, use_payload=False,
                csrf_token=None
            )
```

vojtechjelinek replied:
Done.

aasiffaizal commented on core/controllers/android_test.py:
```diff
@@ -0,0 +1,320 @@
+
+    def setUp(self) -> None:
+        super().setUp()
+        self.secrets_swap = self.swap_to_always_return(
+            secrets_services, 'get_secret', 'secret')
```
```suggestion
        self.secrets_swap = self.swap_to_always_return(
            secrets_services, 'get_secret', 'secret'
        )
```

vojtechjelinek replied:
Done.

aasiffaizal commented on core/controllers/android_test.py:
```diff
@@ -0,0 +1,320 @@
+            secrets_services, 'get_secret', 'secret')
+
+    def test_get_with_wrong_api_key_returns_error(self) -> None:
+        secrets_swap = self.swap_to_always_return(
+            secrets_services, 'get_secret', 'not_key')
```
```suggestion
        secrets_swap = self.swap_to_always_return(
            secrets_services, 'get_secret', 'not_key'
        )
```

vojtechjelinek replied:
Done.

aasiffaizal reviewed (approved):
A few nits. Otherwise LGTM for code owner files.

BenHenning commented on core/controllers/android_test.py:
```diff
@@ -0,0 +1,320 @@
+    models.Names.TRANSLATION])
+
+
+class InitializeAndroidTestDataHandlerTest(test_utils.GenericTestBase):
```
Would be nice to also include a test that verifies calling it twice correctly regenerates a new topic & removes the old one (since this was functionality added to make debugging the new Android data handler easier).

vojtechjelinek replied:
Done.

BenHenning reviewed (commented):
Thanks @vojtechjelinek! I've reviewed my existing & new codeowner files. I've also updated https://github.com/oppia/oppia-android/pull/4885 & verified that the latest batch changes seem to be working correctly (though I can't test the batching bit in detail without production data, so that part will need to come after this PR goes in).

Had a few follow-up comments (plus my old ones that are still open), otherwise the PR looks good.

vojtechjelinek committed:
Address comments

vojtechjelinek committed:
Fix tests

BenHenning commented on scripts/servers.py:
```diff
@@ -158,8 +158,10 @@ def managed_dev_appserver(
             structure of the server.
-        env: dict(str: str) or None. Defines the environment variables for the
-            new process.
+        env: dict(str: str) or None. Defines additional environment
```
Docstring may no longer be accurate since it seems the merging part was reverted. You may want to consider keeping that behavior since it's more correct.

BenHenning commented on scripts/run_e2e_tests.py:
```diff
@@ -219,7 +219,6 @@ def run_tests(args: argparse.Namespace) -> Tuple[List[bytes], int]:
             automatic_restart=False,
             skip_sdk_update_check=True,
             env={
-                **os.environ,
```
This is now a discrepancy since the environment variable merging is no longer happening in servers.py.

BenHenning reviewed (approved):
Thanks @vojtechjelinek! PR LGTM from Android perspective. I found a few extra potential issues that are outside the Android-specific fixes, though--PTAL.

JayVivarekar reviewed (approved):
LGTM for codeowner files

vojtechjelinek committed:
Merge branch 'develop' of https://github.com/oppia/oppia into add-new-android-handler

vojtechjelinek committed:
Address comments

vojtechjelinek committed:
Merge branch 'develop' into add-new-android-handler

kevintab95 reviewed (approved):
LGTM, thanks @vojtechjelinek!

vojtechjelinek committed:
Merge branch 'add-new-android-handler' of https://github.com/vojtechjelinek/oppia into add-new-android-handler

vojtechjelinek commented:
@seanlip Can you please approve.

seanlip commented on core/controllers/android_test.py:
```diff
@@ -0,0 +1,349 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for the android."""
```
Tests for the android handler.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/classroom_config_services.py:
```diff
@@ -35,6 +35,12 @@
 
 (classroom_models,) = models.Registry.import_models([models.Names.CLASSROOM])
 
+# TODO(#17246): Currently, the classroom data is stored in the config model and
```
@vojtechjelinek Why rename as classroom_domain, shouldn't it be classroom_services (and ditto for the rest of that comment)?

vojtechjelinek replied:
Done.

seanlip commented on core/domain/classroom_domain.py:
```diff
@@ -18,7 +18,23 @@
+    course_details: str
+    topic_list_intro: str
+
+# TODO(#17246): Currently, the classroom data is stored in the config model and
```
@vojtechjelinek Not sure this makes sense, this file is already classroom_domain.

vojtechjelinek replied:
Done.

seanlip commented on core/domain/classroom_services.py:
```diff
@@ -24,6 +24,12 @@
 
+# TODO(#17246): Currently, the classroom data is stored in the config model and
+# we are planning to migrate the storage into a new Classroom model. After the
+# successful migration, this file should be renamed as classroom_domain and
```
Ditto, not sure this makes sense, don't we need to do something else with this file rather than rename it?

vojtechjelinek replied:
Done.

seanlip reviewed (changes requested):
Thanks @vojtechjelinek, took a pass, just minor comments.

vojtechjelinek committed:
Merge branch 'develop' into add-new-android-handler

vojtechjelinek committed:
Update android_test.py

vojtechjelinek committed:
Update classroom_config_services.py

vojtechjelinek committed:
Update classroom_domain.py

vojtechjelinek committed:
Update classroom_services.py

seanlip commented on core/domain/classroom_config_services.py:
```diff
@@ -37,8 +37,8 @@
-# successful migration, this file should be renamed as classroom_domain and
-# the exiting classroom domain file should be deleted, until then both of
+# successful migration, this file should be renamed as classroom_services and
+# the exiting classroom services file should be deleted, until then both of
```
exiting --> existing

seanlip reviewed (approved):
Thanks @vojtechjelinek, LGTM. Just one typo fix.

vojtechjelinek committed:
Update classroom_config_services.py
